core/src/main/java/org/elasticsearch/index/query/GeoDistanceRangeQueryBuilder.java

```diff
@@ -111,65 +146,295 @@ public GeoDistanceRangeQueryBuilder includeLower(boolean includeLower) {
+        if (geohash == null) {
+            point = new GeoPoint(lat, lon);
+        } else {
+            point = new GeoPoint();
+            GeoHashUtils.decode(geohash, point);
+        }
+        // validation was not available prior to 2.x, so to support bwc
+        // percolation queries we only ignore_malformed on 2.x created indexes
+        if (!indexCreatedBeforeV2_0 && !ignoreMalformed) {
+            if (point.lat() > 90.0 || point.lat() < -90.0) {
+                throw new QueryParsingException(context.parseContext(), "illegal latitude value [{}] for [{}]", point.lat(), NAME);
+            }
+            if (point.lon() > 180.0 || point.lon() < -180) {
+                throw new QueryParsingException(context.parseContext(), "illegal longitude value [{}] for [{}]", point.lon(), NAME);
+            }
+        }
+
+        if (coerce) {
+            GeoUtils.normalizePoint(point, coerce, coerce);
+        }
+
+        Double fromValue = null;
+        Double toValue = null;
+        if (from != null) {
+            if (from instanceof Number) {
+                fromValue = unit.toMeters(((Number) from).doubleValue());
+            } else {
+                fromValue = DistanceUnit.parse((String) from, unit, DistanceUnit.DEFAULT);
+            }
+            fromValue = geoDistance.normalize(fromValue, DistanceUnit.DEFAULT);
+        }
+        if (to != null) {
+            if (to instanceof Number) {
+                toValue = unit.toMeters(((Number) to).doubleValue());
+            } else {
+                toValue = DistanceUnit.parse((String) to, unit, DistanceUnit.DEFAULT);
+            }
+            toValue = geoDistance.normalize(toValue, DistanceUnit.DEFAULT);
+        }
+
+        MappedFieldType fieldType = context.fieldMapper(fieldName);
+        if (fieldType == null) {
+            throw new QueryParsingException(context.parseContext(), "failed to find geo_point field [" + fieldName + "]");
+        }
+        if (!(fieldType instanceof GeoPointFieldMapper.GeoPointFieldType)) {
+            throw new QueryParsingException(context.parseContext(), "field [" + fieldName + "] is not a geo_point field");
+        }
+        GeoPointFieldMapper.GeoPointFieldType geoFieldType = ((GeoPointFieldMapper.GeoPointFieldType) fieldType);
+
+        IndexGeoPointFieldData indexFieldData = context.getForField(fieldType);
+        Query query = new GeoDistanceRangeQuery(point, fromValue, toValue, includeLower, includeUpper, geoDistance, geoFieldType,
+                indexFieldData, optimizeBbox);
+        if (queryName != null) {
+            context.addNamedQuery(queryName, query);
+        }
+        return query;
+    }
+
     @Override
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(NAME);
         if (geohash != null) {
-            builder.field(name, geohash);
+            builder.field(fieldName + GeoPointFieldMapper.Names.GEOHASH_SUFFIX, geohash);
```

Comment:
So the parsing of the geoHash/point is tricky here as we have the following logic:

* if the field is FIELDNAME.geohash, parse it as a geohash string
* if the field is FIELDNAME.lat parse it as a latitude number
* if the field is FIELDNAME.lon parse it as a longitude number
* if the field is FIELDNAME pass it to GeoPoint.fromString() which will work out if its a geohash or a point string and parse it appropriately but return a GeoPoint.

The problem is that we need to explicitly define if its a geohash when we render so that we can make sure that we parse and render the same way (i.e. so that geohashes we render are parsed back into the geohash field in the builder.

Maybe a solution here would be to remove the geohash field form the builder and only have point, then the geohash(String) method would call GeoPoint.fromString() to convert the geohash to a point?